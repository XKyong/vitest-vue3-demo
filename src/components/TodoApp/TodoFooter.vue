<template>
  <footer class="footer">
    <!-- This should be `0 items left` by default -->
    <span class="todo-count"
      ><strong data-testid="done-todos-count">{{ doneTodosCount }}</strong> item
      left</span
    >
    <!-- Remove this if you don't implement routing -->
    <ul class="filters">
      <li>
        <RouterLink to="/">All</RouterLink>
      </li>
      <li>
        <RouterLink to="/active">Active</RouterLink>
      </li>
      <li>
        <RouterLink to="/completed">Completed</RouterLink>
      </li>
    </ul>
    <!-- Hidden if no completed items are left â†“ -->
    <button
      v-if="clearBtnVisible"
      data-testid="clear-completed"
      class="clear-completed"
      @click="$emit('clear-completed')"
    >
      Clear completed
    </button>
  </footer>
</template>
<script setup lang="ts">
import { computed } from 'vue'

interface TodoItem {
  id: number
  text: string
  done: boolean
}

interface Props {
  todos: TodoItem[]
}

const props = withDefaults(defineProps<Props>(), {
  todos: () => [] as TodoItem[],
})

const emit = defineEmits(['clear-completed'])

const doneTodosCount = computed(
  () => props.todos.filter((todo) => !todo.done).length
)

const clearBtnVisible = computed(() => props.todos.some((todo) => todo.done))
</script>
