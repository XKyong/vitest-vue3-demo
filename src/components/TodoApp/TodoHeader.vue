<template>
  <header class="header">
    <h1>todos</h1>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
      autofocus
      data-testid="new-todo"
      @keyup.enter="handleNewTodo"
    />
  </header>
</template>
<script setup lang="ts">

const emit = defineEmits(['new-todo'])

function handleNewTodo(e: KeyboardEvent) {
  const inputEl = e.target as HTMLInputElement
  const value = inputEl.value.trim()
  if (!value.length) return
  emit('new-todo', value)
  inputEl.value = ''
}
</script>
