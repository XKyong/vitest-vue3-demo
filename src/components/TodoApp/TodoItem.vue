<template>
  <li data-testid="todo-item" :class="{ 'completed': todo.done, 'editing': isEditing }">
    <div class="view">
      <input v-model="todo.done" data-testid="todo-done" class="toggle" type="checkbox" />
      <label data-testid="todo-text" @dblclick="isEditing = true">{{ todo.text }}</label>
      <button data-testid="delete" class="destroy" @click="$emit('delete-todo', todo.id)"></button>
    </div>
    <input v-focus="todo.text" data-testid="todo-edit" class="edit" :value="todo.text" @blur="isEditing = false" @keyup.enter="handleEditTodo" />
  </li>
</template>
<script setup lang="ts">
import { ref, type DirectiveBinding } from 'vue'
interface TodoItem {
  id: number
  text: string
  done: boolean
}

interface Props {
  todo: TodoItem
}

const props = withDefaults(defineProps<Props>(), {
  todo: () => ({} as TodoItem),
})

const emit = defineEmits<{
  (e: 'delete-todo', id: number): void
  (e: 'edit-todo', data: { id: number; text: string }): void
}>()

const isEditing = ref(false)

const vFocus = (el: HTMLInputElement, binding: DirectiveBinding<string>) => {
  if (binding.value) {
    el.focus()
  }
}

function handleEditTodo(e: KeyboardEvent) {
  emit('edit-todo', {
    id: props.todo.id,
    text: (e.target as HTMLInputElement).value
  })
}
</script>
