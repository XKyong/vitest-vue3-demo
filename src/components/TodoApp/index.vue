<template>
  <section class="todoapp">
    <TodoHeader @new-todo="handleNewTodo" />
    <!-- This section should be hidden by default and shown when there are todos -->
    <section class="main">
      <input
        v-model="toggleAll"
        data-testid="toggle-all"
        id="toggle-all"
        class="toggle-all"
        type="checkbox"
      />
      <label for="toggle-all">Mark all as complete</label>
      <ul class="todo-list">
        <!-- These are here just to show the structure of the list items -->
        <!-- List items should get the class `editing` when editing and `completed` when marked as completed -->
        <TodoItem
          v-for="todo in todos"
          :key="todo.id"
          :todo="todo"
          @delete-todo="handleDeleteTodo"
          @edit-todo="handleEditTodo"
        />
      </ul>
    </section>
    <!-- This footer should be hidden by default and shown when there are todos -->
    <TodoFooter />
  </section>
</template>
<script setup lang="ts">
import { ref, computed } from 'vue'
import TodoHeader from './TodoHeader.vue'
import TodoFooter from './TodoFooter.vue'
import TodoItem from './TodoItem.vue'

interface TodoItem {
  id: number
  text: string
  done: boolean
}

const todos = ref<TodoItem[]>([])

const toggleAll = computed({
  get() {
    return !!todos.value.length && todos.value.every((todo) => todo.done)
  },
  set(checked: boolean) {
    todos.value.forEach((todo) => {
      todo.done = checked
    })
  }
})

function handleNewTodo(text: string) {
  const endTodo = todos.value[todos.value.length - 1]
  todos.value.push({
    id: endTodo ? endTodo.id + 1 : 1,
    text,
    done: false,
  })
}

function handleDeleteTodo(id: number) {
  todos.value = todos.value.filter((todo) => todo.id !== id)
}

function handleEditTodo(todo: { id: number; text: string }) {
  const index = todos.value.findIndex((item) => item.id === todo.id)
  if (index === -1) return
  if (todo.text.trim().length === 0) {
    todos.value.splice(index, 1)
  } else {
    todos.value[index].text = todo.text
  }
}

// expose 出去，这样 test 单元测试文件才可以拿到对应方法和类型
defineExpose({
  todos,
  handleNewTodo,
  handleDeleteTodo,
  handleEditTodo,
})
</script>
