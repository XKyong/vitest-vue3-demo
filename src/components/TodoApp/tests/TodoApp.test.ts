import { shallowMount } from '@vue/test-utils'
import { nextTick } from 'vue'
import TodoApp from '../index.vue'  
import TodoItem from '../TodoItem.vue'

import type { VueWrapper } from '@vue/test-utils'

describe('TodoApp.vue', () => {
  let wrapper: VueWrapper | null = null
  beforeEach(() => {
    wrapper = shallowMount(TodoApp)
    const vm = wrapper.vm as InstanceType<typeof TodoApp>
    vm.todos = [
      { id: 1, text: 'eat', done: false },
      { id: 2, text: 'play', done: false },
      { id: 3, text: 'sleep', done: false }
    ]
  })

  it('New todo', async () => {
    const vm = wrapper!.vm as InstanceType<typeof TodoApp>
    const text = 'play'
    vm.handleNewTodo(text)
    const todo = vm.todos.find(t => t.text === text)
    expect(todo).toBeTruthy()
  })

  it('Todo List', async () => {
    const vm = wrapper!.vm as InstanceType<typeof TodoApp>
    // `wrapper.setData` will emit error here: `TypeError: Cannot add property todos, object is not extensible`
    // await wrapper.setData({ todos })
    await nextTick()
    expect(wrapper!.findAllComponents(TodoItem).length).toBe(vm.todos.length)
  })

  it('Delete todo', async () => {
    const vm = wrapper!.vm as InstanceType<typeof TodoApp>
    await vm.handleDeleteTodo(1)
    expect(vm.todos.length).toBe(2)
    expect(wrapper!.findAllComponents(TodoItem).length).toBe(2)
  })

  it('Delete not existed todo', async () => {
    const vm = wrapper!.vm as InstanceType<typeof TodoApp>
    await vm.handleDeleteTodo(100)
    expect(vm.todos.length).toBe(3)
    expect(wrapper!.findAllComponents(TodoItem).length).toBe(3)
  })

  it('Edit Todo', async () => {
    const todo = { id: 2, text: 'abc' }
    const vm = wrapper!.vm as InstanceType<typeof TodoApp>
    await vm.handleEditTodo(todo)
    expect(vm.todos[1].text).toBe(todo.text)

    // 当我传入的 text 是空字符串时，执行删除操作
    todo.text = ''
    await vm.handleEditTodo(todo)
    expect(vm.todos.find(t => t.id === todo.id)).toBeFalsy()
  })

  it('Toggle all', async () => {
    const vm = wrapper!.vm as InstanceType<typeof TodoApp>
    const toggleAll = wrapper!.find('[data-testid="toggle-all"]')
    // 断言所有的子任务都被选中
    toggleAll.setValue(true)
    expect(vm.todos.every(t => t.done)).toBeTruthy()
    // 断言所有的子任务都不被选中
    toggleAll.setValue(false)
    expect(vm.todos.every(t => !t.done)).toBeTruthy()
  })

  it('Toggle all state', async () => {
    const toggleAll = wrapper!.find('[data-testid="toggle-all"]')
    const vm = wrapper!.vm as InstanceType<typeof TodoApp>

    // 当所有任务都被选中时，toggleAll 应该被选中
    vm.todos.forEach(todo => (todo.done = true))
    await nextTick()
    expect((toggleAll.element as HTMLInputElement).checked).toBeTruthy()

    // 当有一个任务未被选中时，toggleAll 应该未被选中
    vm.todos[0].done = false
    await nextTick()
    expect((toggleAll.element as HTMLInputElement).checked).toBeFalsy()

  })
})
