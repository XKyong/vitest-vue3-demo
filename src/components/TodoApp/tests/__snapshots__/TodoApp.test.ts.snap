// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoApp.vue > snapshot 1`] = `
"<section class="todoapp">
  <todo-header-stub></todo-header-stub><!-- This section should be hidden by default and shown when there are todos -->
  <section class="main"><input data-testid="toggle-all" id="toggle-all" class="toggle-all" type="checkbox"><label for="toggle-all">Mark all as complete</label>
    <ul class="todo-list">
      <!-- These are here just to show the structure of the list items -->
      <!-- List items should get the class \`editing\` when editing and \`completed\` when marked as completed -->
      <todo-item-stub todo="[object Object]"></todo-item-stub>
    </ul>
  </section><!-- This footer should be hidden by default and shown when there are todos -->
  <todo-footer-stub todos="[object Object],[object Object],[object Object]"></todo-footer-stub>
</section>"
`;
